tket/src/Mapping/MappingFrontier.cpp

```diff
@@ -257,160 +260,196 @@ void MappingFrontier::advance_frontier_boundary(
+     * Classical and Boolean Edges, we find a "next" set
+     * of causally forward adjacent vertices.
+     */
     boundary_updated = false;
     std::shared_ptr<unit_frontier_t> l_frontier_edges =
         frontier_convert_vertport_to_edge(
             this->circuit_, this->linear_boundary);
     CutFrontier next_cut =
         this->circuit_.next_cut(l_frontier_edges, this->boolean_boundary);
-    // For each vertex in a slice, if its physically permitted, update
-    // linear_boundary with quantum out edges from vertex (i.e.
-    // next_cut.u_frontier)
-    // update boolean_boundary in line
+
+    /**
+     * For each vertex in a slice we check to see if its
+     * Quantum arguments are permitted by the Architecture.
+     *
+     * If true, we update our Circuit Slice by replacing
+     * "in edges" to the Vertex with appropriate "out edges"
+     * from the Vertex.
+     */
     for (const Vertex& vert : *next_cut.slice) {
-      // for each boolean edge into vertex, collect associated Bit and port
-      // number n.b. a single Bit may have multiple "in bundles" to different
-      // vertices in the same cut
-      std::map<Bit, port_t> bool_uid_port_set;
-      std::vector<EdgeVec> b_in_bundles = this->circuit_.get_b_in_bundles(vert);
-      for (unsigned i = 0; i < b_in_bundles.size(); i++) {
-        EdgeVec ev = b_in_bundles[i];
-        if (ev.size() > 0) {
-          bool_uid_port_set.insert(
-              {get_bit_from_bool_frontier(this->boolean_boundary, ev), i});
+      /**
+       * Iterate through every edge (port ordered) to the Vertex.
+       * For each edge, store it's associated UnitID and Edge Type.
+       */
+      std::vector<std::pair<UnitID, EdgeType>> in_uids;
+      std::vector<Node> nodes;
+      std::vector<Edge> all_in_edges = this->circuit_.get_in_edges(vert);
+      for (const Edge& edge : all_in_edges) {
+        UnitID uid;
+        EdgeType edge_type = this->circuit_.get_edgetype(edge);
+        switch (edge_type) {
+          case EdgeType::Quantum: {
+            uid = get_unitid_from_unit_frontier(
+                this->linear_boundary, {this->circuit_.source(edge),
+                                        this->circuit_.get_source_port(edge)});
+            // We use "nodes" to check if Quantum arguments respect the
+            // architecture
+            nodes.push_back(Node(uid));
+            break;
+          }
+          case EdgeType::Classical: {
+            uid = get_unitid_from_unit_frontier(
+                this->linear_boundary, {this->circuit_.source(edge),
+                                        this->circuit_.get_source_port(edge)});
+            break;
+          }
+          case EdgeType::Boolean: {
+            // N.B. function will find uid even if the bundle has more than the
+            // single "edge" passed
+            uid = get_bit_from_bool_frontier(this->boolean_boundary, {edge});
+            break;
+          }
```

Comment:
Worth having a `dafault` case with a `TKET_ASSERT` that aborts. It's likely that a new edge type (WASM) will be added in the future and we want to make sure we catch this.

Reply:
Done